File: labs/_lib/provenance.R
# ---------------------------------------------------------------------------
# provenance.R — make a build script notice when its source moves.
#
# WHY THIS EXISTS
#
# The campaign-visits lab pinned an Associated Press dataset at version /50/
# and captured 323 events. By the time anyone looked again the AP was serving
# /54/ with 376 events — the committed file stopped three days before the
# election and was missing the entire closing push. Nothing failed. Nothing
# warned. The build script ran clean and produced stale data.
#
# That is the failure mode this file exists to catch: a URL that still returns
# 200, still parses, and no longer means what it meant.
#
# WHAT IT DOES
#
# Wrap a download and it records, in `PROVENANCE.tsv` next to the data:
#
#     url, first_seen, last_seen, bytes, sha256, rows
#
# On every later run it compares. If the bytes, hash or row count changed it
# prints a loud banner saying WHAT changed and BY HOW MUCH — then carries on,
# because a source legitimately updating is normal and should not break a
# build. The point is that a human sees it.
#
# USE
#
#     source("../../../_lib/provenance.R")       # from a lab's data/ folder
#     prov_fetch(URL, "raw.csv")                 # instead of download.file()
#     d <- prov_read_csv(URL)                    # instead of read.csv(URL)
#     prov_report()                              # at the end of the script
#     prov_stamp()                               # at the end of the script
#
# `prov_fetch` returns the destination path, so it drops into existing code:
#     download.file(u, f)      ->  prov_fetch(u, f)
#     read.csv(u)              ->  prov_read_csv(u)
# ---------------------------------------------------------------------------

PROV_FILE <- "PROVENANCE.tsv"
.prov_changes <- new.env(parent = emptyenv())
.prov_changes$rows <- list()

# When the R process started -- NOT when this file was sourced. prov_stamp()
# uses it to tell what this run wrote from what was already lying in the
# directory, and a script that sources the helper on its last line rather than
# its first would otherwise appear to have produced nothing at all.
# proc.time()["elapsed"] is wall-clock seconds since the session began.
PROV_START <- Sys.time() - as.numeric(proc.time()[["elapsed"]])

.prov_load <- function() {
  if (!file.exists(PROV_FILE)) {
    return(data.frame(url = character(), first_seen = character(),
                      last_seen = character(), bytes = numeric(),
                      sha256 = character(), rows = character(),
                      stringsAsFactors = FALSE))
  }
  read.delim(PROV_FILE, stringsAsFactors = FALSE, colClasses = "character")
}

.prov_sha <- function(path) {
  # tools::md5sum is base; prefer sha256 via openssl if available, else md5.
  #
  # THE CONNECTION MUST BE OPENED "rb". file(path) opens in TEXT mode, and on a
  # binary source -- an .RData, a .dta, a .zip -- openssl then digests something
  # no other tool reproduces. A hash nobody else can compute is not a hash: it
  # cannot be checked against the publisher, and `stamp-builds.py`, which reads
  # raw bytes, reports every such file as "content changed" forever after.
  # Caught on media-ideology/raw/PolShares.RData, where text mode gave
  # 1d734c32... and the true digest is fa320555...  CSV captures were unaffected,
  # which is why this survived: every earlier raw capture in the corpus is text.
  # THE CONNECTION MUST ALSO BE CLOSED. openssl reads it to the end but leaves
  # it open, so a bare sha256(file(path, "rb")) leaks one connection per file
  # and the garbage collector later prints "closing unused connection N" for
  # every one of them -- a build that stamps 11 files ends in 11 warnings that
  # look like they belong to the build. Streamed rather than read whole,
  # because raw captures are kept at whatever size they arrived.
  if (requireNamespace("openssl", quietly = TRUE)) {
    con <- file(path, "rb")
    on.exit(close(con), add = TRUE)
    return(paste0("sha256:", as.character(openssl::sha256(con))))
  }
  paste0("md5:", unname(tools::md5sum(path)))
}

.prov_rows <- function(path) {
  # Cheap row estimate. Not meaningful for binary formats; returns NA there.
  ext <- tolower(tools::file_ext(path))
  if (!ext %in% c("csv", "tsv", "tab", "txt", "json", "xml")) return(NA_character_)
  n <- tryCatch(length(readLines(path, warn = FALSE)), error = function(e) NA_integer_)
  if (is.na(n)) return(NA_character_)
  as.character(if (ext %in% c("csv", "tsv", "tab")) max(n - 1L, 0L) else n)
}

# Strip credentials before anything is written to disk. PROVENANCE.tsv is a
# committed file, so a key in a query string would be published with the lab.
# This happened once: an ACS build wrote a live Census key into 22 rows.
.prov_scrub <- function(url) {
  gsub("([?&](key|api_key|apikey|token|access_token|signature|secret)=)[^&#]*",
       "\\1REDACTED", url, ignore.case = TRUE)
}

.prov_record <- function(url, path, label = NULL) {
  url <- .prov_scrub(url)
  p   <- .prov_load()
  now <- format(Sys.Date())
  new <- list(url = url, bytes = as.character(file.size(path)),
              sha256 = .prov_sha(path), rows = .prov_rows(path))
  hit <- which(p$url == url)

  if (length(hit) == 0) {
    p <- rbind(p, data.frame(url = url, first_seen = now, last_seen = now,
                             bytes = new$bytes, sha256 = new$sha256,
                             rows = new$rows, stringsAsFactors = FALSE))
    cat(sprintf("  [prov] first capture: %s  (%s bytes%s)\n",
                basename(path), new$bytes,
                if (is.na(new$rows)) "" else paste0(", ", new$rows, " rows")))
  } else {
    old <- p[hit[1], ]
    moved <- c(
      if (!identical(old$bytes,  new$bytes))  sprintf("bytes %s -> %s", old$bytes,  new$bytes),
      if (!identical(old$sha256, new$sha256)) "content hash changed",
      if (!identical(old$rows,   new$rows) && !is.na(new$rows))
        sprintf("rows %s -> %s (%+d)", old$rows, new$rows,
                suppressWarnings(as.integer(new$rows) - as.integer(old$rows))))
    if (length(moved)) {
      .prov_changes$rows[[length(.prov_changes$rows) + 1]] <-
        list(url = url, file = basename(path), first = old$first_seen,
             moved = moved, label = label)
    }
    p[hit[1], c("last_seen", "bytes", "sha256", "rows")] <-
      list(now, new$bytes, new$sha256, new$rows)
  }
  write.table(p, PROV_FILE, sep = "\t", row.names = FALSE, quote = FALSE)
  invisible(path)
}

#' Download a URL and record its provenance.
prov_fetch <- function(url, dest, label = NULL, mode = "wb", quiet = TRUE, ...) {
  utils::download.file(url, dest, mode = mode, quiet = quiet, ...)
  .prov_record(url, dest, label)
  # Invisible, like download.file(). A bare prov_fetch() at the top of a script
  # would otherwise auto-print its return value, and every build log would
  # carry a line of [1] "raw/whatever.zip" for each file it collected.
  invisible(dest)
}

#' read.csv() from a URL, with provenance recorded.
prov_read_csv <- function(url, label = NULL, ...) {
  tmp <- tempfile(fileext = paste0(".", tools::file_ext(sub("\\?.*$", "", url))))
  utils::download.file(url, tmp, mode = "wb", quiet = TRUE)
  .prov_record(url, tmp, label)
  utils::read.csv(tmp, ...)
}

# ---------------------------------------------------------------------------
# THE BUILD STAMP
#
# PROVENANCE.tsv, above, answers "where did this file come from, and has the
# source moved since". It can only answer that for a file that was downloaded,
# which leaves out about half the corpus: a chapter whose data is simulated,
# derived from another chapter, or typed out of a published table fetches
# nothing, so it records nothing, and the book cannot say anything about it.
#
# BUILD-STAMP.tsv answers the smaller question every chapter can answer --
# which script produced what is sitting in this directory, when it last ran,
# and what the files looked like when it finished:
#
#     script  stamped_on  stamp_source  file  bytes  sha256  rows  file_mtime
#
# `stamp_source` is the honest column. "build" means a running script wrote
# the row and `stamped_on` is a real run date. "disk" means the row was taken
# by reading the directory afterwards -- the sizes and hashes are real, the
# build date is not known, and file_mtime is the best evidence there is. A
# backfilled chapter promotes itself to "build" the next time its script runs.
#
# ONLY WHAT THIS RUN WROTE. Files whose mtime predates PROV_START are left
# with whatever row they already had, rather than being restamped with today's
# date. A build that deliberately reuses a cached download should not be able
# to claim it fetched it this morning, and that is the whole point of the file.
# ---------------------------------------------------------------------------

STAMP_FILE <- "BUILD-STAMP.tsv"

STAMP_COLS <- c("script", "stamped_on", "stamp_source", "file",
                "bytes", "sha256", "rows", "file_mtime")

.stamp_empty <- function() {
  d <- as.data.frame(setNames(rep(list(character()), length(STAMP_COLS)), STAMP_COLS),
                     stringsAsFactors = FALSE)
  d
}

.stamp_load <- function() {
  if (!file.exists(STAMP_FILE)) return(.stamp_empty())
  read.delim(STAMP_FILE, stringsAsFactors = FALSE, colClasses = "character")
}

# Which script is running. Rscript puts it on the command line; source() leaves
# it on the frame stack. R encodes spaces in --file= as "~+~", which matters
# here because these trees live under "My Drive".
.stamp_script <- function() {
  a <- grep("^--file=", commandArgs(), value = TRUE)
  if (length(a)) {
    return(basename(gsub("~+~", " ", sub("^--file=", "", a[1]), fixed = TRUE)))
  }
  for (i in rev(seq_len(sys.nframe()))) {
    of <- sys.frame(i)$ofile
    if (!is.null(of)) {
      b <- basename(as.character(of)[1])
      if (b != "provenance.R") return(b)
    }
  }
  NA_character_
}

# Everything a build is allowed to have produced. Data lives in derived/ and
# raw/ by convention (check-layout.py enforces it), so a file at the top of
# data/ is metadata or a script and is not a build output.
.stamp_outputs <- function() {
  fs <- character()
  for (d in c("derived", "raw")) {
    if (dir.exists(d)) fs <- c(fs, list.files(d, recursive = TRUE, full.names = TRUE))
  }
  fs[!dir.exists(fs)]
}

#' Record which script built what is in this directory, and when.
#'
#' @param outputs  files to stamp. The default, NULL, is everything under
#'                 derived/ and raw/ that THIS RUN wrote. "all" is everything
#'                 there whenever it was written -- for a backfill, or for a
#'                 script whose outputs cannot be told apart by their mtime.
#' @param script   name to record; default is the running script
#' @param source_kind  "build" from a running script, "disk" when backfilling
prov_stamp <- function(outputs = NULL, script = NULL, source_kind = "build") {
  script <- if (is.null(script)) .stamp_script() else script
  all_out <- .stamp_outputs()
  fs <- if (identical(outputs, "all")) all_out
        else if (!is.null(outputs)) outputs
        else {
          info <- file.info(all_out)
          all_out[!is.na(info$mtime) & info$mtime >= PROV_START]
        }

  p <- .stamp_load()
  # Drop rows for files that are no longer there. A stamp that still lists a
  # deleted file is worse than no stamp: it reads as a promise.
  if (nrow(p)) p <- p[p$file %in% all_out, , drop = FALSE]

  if (!length(fs)) {
    write.table(p, STAMP_FILE, sep = "\t", row.names = FALSE, quote = FALSE)
    cat("  [stamp] no new outputs this run;", nrow(p), "row(s) kept\n")
    return(invisible(p))
  }

  now  <- format(Sys.Date())
  info <- file.info(fs)
  new  <- data.frame(
    script       = script,
    stamped_on   = now,
    stamp_source = source_kind,
    file         = fs,
    bytes        = as.character(info$size),
    sha256       = vapply(fs, .prov_sha, character(1), USE.NAMES = FALSE),
    rows         = vapply(fs, .prov_rows, character(1), USE.NAMES = FALSE),
    # format(), not as.Date(): as.Date() on a POSIXct converts in UTC unless
    # told otherwise, which stamps a file written this evening with tomorrow's
    # date and makes file_mtime disagree with stamped_on for no visible reason.
    file_mtime   = format(info$mtime, "%Y-%m-%d"),
    stringsAsFactors = FALSE)

  p <- rbind(p[!p$file %in% new$file, , drop = FALSE], new)
  p <- p[order(p$file), , drop = FALSE]
  write.table(p, STAMP_FILE, sep = "\t", row.names = FALSE, quote = FALSE)
  cat(sprintf("  [stamp] %s: %d file(s) %s %s\n", script, nrow(new),
              if (identical(source_kind, "build")) "built on" else "read off disk on",
              now))
  invisible(p)
}

#' Print the drift banner. Call once at the end of a build script.
prov_report <- function() {
  ch <- .prov_changes$rows
  if (!length(ch)) {
    cat("  [prov] all sources unchanged since last build\n")
    return(invisible(FALSE))
  }
  bar <- strrep("=", 72)
  cat("\n", bar, "\n", sep = "")
  cat("  *** SOURCE DATA CHANGED SINCE THE LAST BUILD ***\n\n")
  for (c_ in ch) {
    cat(sprintf("  %s\n    %s\n    first captured %s\n",
                c_$file, c_$url, c_$first))
    for (m in c_$moved) cat("      - ", m, "\n", sep = "")
    cat("\n")
  }
  cat("  Every figure derived from these files may have moved. Re-check any\n")
  cat("  number quoted in the lab, the key or the brief before trusting it.\n")
  cat(bar, "\n\n", sep = "")
  invisible(TRUE)
}

File: labs/_lib/provenance.py
"""provenance.py — make a build script notice when its source moves.

WHY THIS EXISTS

The campaign-visits lab pinned an Associated Press dataset at version /50/ and
captured 323 events. By the time anyone looked again the AP was serving /54/
with 376 events -- the committed file stopped three days before the election
and was missing the entire closing push. Nothing failed. Nothing warned. The
build ran clean and produced stale data.

That is the failure mode this module exists to catch: a URL that still returns
200, still parses, and no longer means what it meant.

WHAT IT DOES

Wrap a download and it records, in `PROVENANCE.tsv` next to the data:

    url, first_seen, last_seen, bytes, sha256, rows

On every later run it compares. If bytes, hash or row count changed it prints a
loud banner saying WHAT changed and BY HOW MUCH, then carries on -- a source
legitimately updating is normal and should not break a build. The point is that
a human sees it.

USE

    import sys; sys.path.insert(0, "../_lib")
    from provenance import fetch, report

    fetch(URL, "raw.csv")          # instead of urllib.request.urlretrieve
    ...
    report()                       # at the end of the script
    stamp()                        # at the end of the script

`fetch` returns the destination path, so it drops into existing code.
"""

import hashlib
import os
import sys
import tempfile
import time
import urllib.parse
import urllib.request
from datetime import date

PROV_FILE = "PROVENANCE.tsv"
COLS = ["url", "first_seen", "last_seen", "bytes", "sha256", "rows"]
_changes = []

# When this module was imported, which is as near as anything gets to when the
# build started. stamp() uses it to tell what the run actually wrote from what
# was already lying in the directory -- see the note above stamp().
START = time.time()

STAMP_FILE = "BUILD-STAMP.tsv"
STAMP_COLS = ["script", "stamped_on", "stamp_source", "file",
              "bytes", "sha256", "rows", "file_mtime"]

_TEXTY = {".csv", ".tsv", ".tab", ".txt", ".json", ".xml"}


def _load():
    if not os.path.exists(PROV_FILE):
        return {}
    out = {}
    with open(PROV_FILE, encoding="utf-8") as fh:
        head = fh.readline().rstrip("\n").split("\t")
        for line in fh:
            v = line.rstrip("\n").split("\t")
            if len(v) == len(head):
                r = dict(zip(head, v))
                out[r["url"]] = r
    return out


def _save(recs):
    with open(PROV_FILE, "w", encoding="utf-8") as fh:
        fh.write("\t".join(COLS) + "\n")
        for r in recs.values():
            fh.write("\t".join(str(r.get(c, "")) for c in COLS) + "\n")


def _sha(path):
    h = hashlib.sha256()
    with open(path, "rb") as fh:
        for chunk in iter(lambda: fh.read(1 << 20), b""):
            h.update(chunk)
    return "sha256:" + h.hexdigest()


def _rows(path):
    if os.path.splitext(path)[1].lower() not in _TEXTY:
        return ""
    try:
        with open(path, "rb") as fh:
            n = sum(1 for _ in fh)
    except OSError:
        return ""
    if os.path.splitext(path)[1].lower() in {".csv", ".tsv", ".tab"}:
        n = max(n - 1, 0)
    return str(n)


def record(url, path, label=None):
    """Record (or compare) one already-downloaded file."""
    recs = _load()
    now = date.today().isoformat()
    new = {"bytes": str(os.path.getsize(path)),
           "sha256": _sha(path), "rows": _rows(path)}
    old = recs.get(url)

    if old is None:
        recs[url] = dict(url=url, first_seen=now, last_seen=now, **new)
        extra = f", {new['rows']} rows" if new["rows"] else ""
        print(f"  [prov] first capture: {os.path.basename(path)} "
              f"({new['bytes']} bytes{extra})")
    else:
        moved = []
        if old.get("bytes") != new["bytes"]:
            moved.append(f"bytes {old.get('bytes')} -> {new['bytes']}")
        if old.get("sha256") != new["sha256"]:
            moved.append("content hash changed")
        if new["rows"] and old.get("rows") != new["rows"]:
            try:
                d = int(new["rows"]) - int(old.get("rows") or 0)
                moved.append(f"rows {old.get('rows')} -> {new['rows']} ({d:+d})")
            except ValueError:
                moved.append(f"rows {old.get('rows')} -> {new['rows']}")
        if moved:
            _changes.append(dict(url=url, file=os.path.basename(path),
                                 first=old.get("first_seen", "?"),
                                 moved=moved, label=label))
        old.update(last_seen=now, **new)
        recs[url] = old
    _save(recs)
    return path


def fetch(url, dest, label=None, timeout=600):
    """Download a URL and record its provenance."""
    urllib.request.urlretrieve(url, dest)
    return record(url, dest, label)


def fetch_bytes(url, label=None, timeout=300, headers=None):
    """Read a URL into memory, recording its provenance on the way.

    For a source the script consumes without ever saving it -- a JSON blob
    parsed straight into a dict, a CSV read into rows and thrown away. Those
    are the easiest fetches to leave unrecorded and no less able to move under
    a build, so the bytes are hashed through a temporary file and the row is
    written exactly as though the file had been kept.

    `headers` matters: several of these sources refuse a request with no user
    agent, and passing them through is the difference between a recorded fetch
    and a 403.
    """
    req = urllib.request.Request(url, headers=headers or {})
    with urllib.request.urlopen(req, timeout=timeout) as r:
        blob = r.read()

    # Name the temp file after the URL rather than taking mkstemp's random
    # name. Two reasons: record() decides whether a row count is meaningful by
    # looking at the extension, and the drift banner prints the basename -- so
    # a random name would report a moved source as "tmpg247f4q8.csv".
    name = os.path.basename(urllib.parse.urlparse(url).path) or "download.bin"
    d = tempfile.mkdtemp()
    tmp = os.path.join(d, name)
    try:
        with open(tmp, "wb") as fh:
            fh.write(blob)
        record(url, tmp, label)
    finally:
        os.unlink(tmp)
        os.rmdir(d)
    return blob


# ---------------------------------------------------------------------------
# THE BUILD STAMP.  The R twin of this file carries the full note; in short:
# PROVENANCE.tsv can only describe a file that was downloaded, which leaves out
# about half the corpus, so BUILD-STAMP.tsv records the smaller thing every
# chapter can say -- which script produced what is in this directory, when it
# last ran, and what the files looked like when it finished.
#
#     script  stamped_on  stamp_source  file  bytes  sha256  rows  file_mtime
#
# stamp_source is "build" when a running script wrote the row and stamped_on is
# a real run date, "disk" when the row was read off the directory afterwards
# and the build date is not known. Only files this run wrote are restamped.
# ---------------------------------------------------------------------------


def _stamp_load():
    if not os.path.exists(STAMP_FILE):
        return {}
    out = {}
    with open(STAMP_FILE, encoding="utf-8") as fh:
        head = fh.readline().rstrip("\n").split("\t")
        for line in fh:
            v = line.rstrip("\n").split("\t")
            if len(v) == len(head):
                r = dict(zip(head, v))
                out[r["file"]] = r
    return out


def _stamp_save(recs):
    with open(STAMP_FILE, "w", encoding="utf-8") as fh:
        fh.write("\t".join(STAMP_COLS) + "\n")
        for k in sorted(recs):
            fh.write("\t".join(str(recs[k].get(c, "")) for c in STAMP_COLS) + "\n")


def _stamp_outputs():
    """Everything a build is allowed to have produced: derived/ and raw/."""
    out = []
    for d in ("derived", "raw"):
        if not os.path.isdir(d):
            continue
        for root, _, files in os.walk(d):
            out.extend(os.path.join(root, f) for f in files)
    return sorted(out)


def stamp(outputs=None, script=None, source_kind="build"):
    """Record which script built what is in this directory, and when."""
    script = script or os.path.basename(sys.argv[0] or "") or ""
    all_out = _stamp_outputs()
    if outputs == "all":
        outputs = all_out
    elif outputs is None:
        outputs = [f for f in all_out if os.path.getmtime(f) >= START]

    recs = _stamp_load()
    # Drop rows for files that are gone. A stamp still listing a deleted file
    # is worse than no stamp: it reads as a promise.
    recs = {k: v for k, v in recs.items() if k in set(all_out)}

    if not outputs:
        _stamp_save(recs)
        print(f"  [stamp] no new outputs this run; {len(recs)} row(s) kept")
        return recs

    now = date.today().isoformat()
    for f in outputs:
        st = os.stat(f)
        recs[f] = {
            "script": script,
            "stamped_on": now,
            "stamp_source": source_kind,
            "file": f,
            "bytes": str(st.st_size),
            "sha256": _sha(f),
            "rows": _rows(f),
            "file_mtime": date.fromtimestamp(st.st_mtime).isoformat(),
        }
    _stamp_save(recs)
    verb = "built on" if source_kind == "build" else "read off disk on"
    print(f"  [stamp] {script}: {len(outputs)} file(s) {verb} {now}")
    return recs


def report():
    """Print the drift banner. Call once at the end of a build script."""
    if not _changes:
        print("  [prov] all sources unchanged since last build")
        return False
    bar = "=" * 72
    print("\n" + bar)
    print("  *** SOURCE DATA CHANGED SINCE THE LAST BUILD ***\n")
    for c in _changes:
        print(f"  {c['file']}\n    {c['url']}\n    first captured {c['first']}")
        for m in c["moved"]:
            print(f"      - {m}")
        print()
    print("  Every figure derived from these files may have moved. Re-check any")
    print("  number quoted in the lab, the key or the brief before trusting it.")
    print(bar + "\n")
    return True
